lang/syn/src/codegen/accounts/constraints.rs

```diff
@@ -682,6 +692,66 @@ fn generate_constraint_associated_token(
     }
 }
 
+fn generate_constraint_token_account(
+    f: &Field,
+    c: &ConstraintTokenAccountGroup,
+) -> proc_macro2::TokenStream {
+    let name = &f.ident;
+    let account_owner = match &c.authority {
+        Some(fa) => quote! { Option::<&anchor_lang::prelude::Pubkey>::Some(&#fa.key()) },
+        None => quote! { Option::<&anchor_lang::prelude::Pubkey>::None },
+    };
+
+    let spl_token_mint = match &c.mint {
+        Some(fa) => quote! { Option::<&anchor_lang::prelude::Pubkey>::Some(&#fa.key()) },
+        None => quote! { Option::<&anchor_lang::prelude::Pubkey>::None },
+    };
+
+    quote! {
+        if #account_owner.is_some() && #name.owner != *#account_owner.unwrap() {
+            return Err(anchor_lang::error::ErrorCode::ConstraintTokenOwner.into());
+        }
+        if #spl_token_mint.is_some() && #name.mint != *#spl_token_mint.unwrap() {
+            return Err(anchor_lang::error::ErrorCode::ConstraintTokenMint.into());
+        }
```

Comment:
Instead of generating code that has extra checks to check if the constraints are given, you can just not generate code if theyre not given!

Example for the owner check:
```rust
let authority_check = match &c.authority {
    Some(authority) => quote! {
        if #name.owner != *#authority {
            return Err(anchor_lang::error::ErrorCode::ConstraintTokenOwner.into());
        }
    },
    None => quote! {}
};

```
and then you return
```rust
quote! {
 #authority_check
 #mint_check
}
```

Reply:
same with the mint check function below

Reply:
Okay, looks good.